docs/madr/2-storing-registry-state.md

```diff
@@ -0,0 +1,107 @@
+# Storing ANT Registry State in AO Process
+
+- Status: proposed
+- Approvers: [Dylan], [Ariel], [Phil]
+- Date: 2024-07-22
+- Authors: [Atticus]
+
+## Context and Problem Statement
+
+There are multiple options for storing state in ao Lua processes, namely thru
+the use of the Lua `table` primitive, and a special build called aos-sqlite
+which exposes the lua-sqlite3 library for using in-memory databases.
+
+Directly using Lua tables for complex data has shown a need for better tooling
+in manipulating and searching data, making the SQLite implementation a potential
+solution. However, its novelty raises concerns about reliability in larger
+products.
+
+## Decision Drivers
+
+The main drivers for this decision are:
+
+- **Scalability**
+  - The storage mechanism should be computation and storage efficient for
+    optimal resource usage and data fetching.
+- **Performance**
+  - Fast queries are desired.
+- **Ecosystem Integration**
+  - Exploring new modules and their potential benefits within the ao ecosystem.
+- **Maintenance**
+  - How many development cycles will be needed to maintain the solution.
+
+## Considered Options
+
+### Option 1: Using Lua Primative Tables
+
+- **Pros:**
+  - Known and stable.
+  - No need for additional testing or integration.
+  - zero overhead.
+- **Cons:**
+  - Demonstrated scalability issues when attempting to retrieve 150k records in
+    ANT's.
+  - Limited performance with larger datasets, namely deeply nested tables.
+  - Lack of strictness on data types.
+  - Lots of if-and-or-buts when interacting with them using certain combinations
+    of data types (for example lists are null terminated on some iterators)
+
+### Option 2: Use aos-sqlite Module
+
+- **Pros:**
+  - Addresses _compute_ scalability issues with direct table usage due to
```

Comment:
Which compute scalability issues? The scalability issue we were concerned about with tables was memory consumption. Computationally it's pretty fast. We're getting getting hash table lookups on addresses.

Reply:
And at the end of the day sqlite needs to build datastructures in their indexes (e.g. b-trees) to facilitate the efficient compute usage during search. So unless you had specific issues with hashtables for our unordered data, it's hard to do too much better.

Reply:
This wasn't meant to say that we have those, just that sqlite compared to tables should be faster to look up the info. No?

Reply:
depends on the schema - a key lookup in a lua table is o(1). you can't get any faster.